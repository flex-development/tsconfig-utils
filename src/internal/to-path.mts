/**
 * @file Internal - toPath
 * @module tsconfig-utils/internal/toPath
 */

import validateURLString from '#internal/validate-url-string'
import {
  ERR_INVALID_URL_SCHEME,
  type ErrInvalidUrlScheme
} from '@flex-development/errnode'
import pathe from '@flex-development/pathe'

export default toPath

/**
 * Convert `input` to a path.
 *
 * > ðŸ‘‰ **Note**: `input` is assumed to be a path if it a string and cannot be
 * > parsed to an URL (checked using {@linkcode pathe.isURL}).
 *
 * @see {@linkcode ErrInvalidUrlScheme}
 *
 * @internal
 *
 * @this {void}
 *
 * @param {URL | string} input
 *  The {@linkcode URL}, URL string, or path to convert
 * @return {string}
 *  `input` as path
 * @throws {ErrInvalidUrlScheme}
 *  If `input` is not a path or `file:` URL
 */
function toPath(
  this: void,
  input: URL | string
): string {
  validateURLString(input, 'input')

  if (typeof input === 'string') {
    if (!pathe.isURL(input)) return pathe.toPosix(input)
    input = new URL(input)
  }

  if (input.protocol === 'file:') return pathe.fileURLToPath(input)
  throw new ERR_INVALID_URL_SCHEME('file')
}
