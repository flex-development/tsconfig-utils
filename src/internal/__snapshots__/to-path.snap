// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:internal/toPath > should return \`input\` as path ("./tsconfig.json") 1`] = `"./tsconfig.json"`;

exports[`unit:internal/toPath > should return \`input\` as path ("file:///tsconfig.json") 1`] = `"/tsconfig.json"`;
